collector/sockstat_linux.go

```diff
@@ -45,78 +42,130 @@ func NewSockStatCollector() (Collector, error) {
 }
 
 func (c *sockStatCollector) Update(ch chan<- prometheus.Metric) error {
-	sockStats, err := getSockStats(procFilePath("net/sockstat"))
+	fs, err := procfs.NewFS(*procPath)
 	if err != nil {
-		return fmt.Errorf("couldn't get sockstats: %s", err)
+		return fmt.Errorf("failed to open procfs: %v", err)
 	}
-	for protocol, protocolStats := range sockStats {
-		for name, value := range protocolStats {
-			v, err := strconv.ParseFloat(value, 64)
-			if err != nil {
-				return fmt.Errorf("invalid value %s in sockstats: %s", value, err)
-			}
-			ch <- prometheus.MustNewConstMetric(
-				prometheus.NewDesc(
-					prometheus.BuildFQName(namespace, sockStatSubsystem, protocol+"_"+name),
-					fmt.Sprintf("Number of %s sockets in state %s.", protocol, name),
-					nil, nil,
-				),
-				prometheus.GaugeValue, v,
-			)
-		}
-	}
-	return err
-}
 
-func getSockStats(fileName string) (map[string]map[string]string, error) {
-	file, err := os.Open(fileName)
+	stat4, err := fs.NetSockstat()
 	if err != nil {
-		return nil, err
+		return fmt.Errorf("failed to get IPv4 sockstat data: %v", err)
```

Comment:
We need the same `os.IsNotExist()` check here in case someone is running v6-only. :grinning: